### DISCUSSION TITLE
App Icon

### DISCUSSION THREAD

**Discussion by gf3** [wezterm/wezterm#2396] [Show and tell]
I've been using WezTerm for the past few days an absolutely loving it. As such I've started hacking on an icon and I wanted to solicit feedback and gauge interest in a little refresh for the WezTerm app icon.

I would love critical feedback or better yet if anyone is willing to iterate on the icon with me I'd be happy to share the [Sketch](https://www.sketch.com) file.

It's a take on a common terminal icon scheme—but with a colourful twist. I like the idea of something bright and welcoming.

Please let me know what you think!

<img src="https://screenshot.pics/s/WezTerm2x.png" width="1280" alt="WezTerm Banner" />

### Without background

<img src="https://screenshot.pics/s/WezTerm_Icon.png" width="512" alt="WezTerm Icon" />

**Comment by gf3** [comment]
I've been iterating on this a little more and playing around with something more skeuomorphic that better represents an actual terminal: a keyboard.

Same idea with the colours and theme but it feels a little more tangible.

Would love some feedback!

### ⌘ + Tab

![Screenshot](https://screenshot.pics/s/Monosnap_2022-08-19_16-40-10.png)

### Icon

![WezTermIcon3](https://user-images.githubusercontent.com/18397/185704412-007ca0a2-8778-496e-b002-51f324bdd6b7.png)

**Reply by gf3** [reply]
I've tweaked the perspective (and material) a tiny bit to give it more depth.

![⌘ tab](https://screenshot.pics/s/Monosnap_2022-08-19_18-12-19.png)


![WezTerm Icon](https://user-images.githubusercontent.com/18397/185713936-90a2ec83-c90d-45be-bb54-f4a8a1ad943a.png)

**Reply by wez** [reply]
Thanks for this!  I do like how this looks, but I am a dark mode vampire and I find the amount of "white" to be too bright for my eyes.  Could you also make a dark mode variant?

I like the purple tones. I'd love it if the Operator Mono `$W` could make it onto that keycap (the current svg should have some outlines that you can lift and import without needing the whole font)

**Reply by gf3** [reply]
@wez thanks for the feedback! here's a dark mode iteration, i think it looks even better:

![WezTerm Icon Dark](https://user-images.githubusercontent.com/18397/185728091-c7d2d330-f1de-42ea-a03d-753a19b532ae.png)

![WezTerm Icon Dark 2](https://user-images.githubusercontent.com/18397/185728096-8a48fdaf-f7a8-45d8-acc1-44c42cd211d4.png)


and i'll try a version with operator mono as well

**Reply by louishuyng** [reply]
Hi @gf3 I found the icon very nice. Can I use it for the WezTerm extension on Raycast? It hasn't been published yet and is still under review at https://github.com/raycast/extensions/pull/25422. 

I would like to ask if you are okay with that?

**Comment by gf3** [comment]
Spent a little more time refining the dark mode key concept.

@wez I tried using Operator Mono as suggested however the SVG from the current icon was too thin—and when I looked into the licensing of Operator Mono it seems that this use case is not covered.

I've made the board inset and I've tweaked the lighting and keycap materials to give it a more polished/shiny effect.

### ⌘ Tab

![⌘ tab](https://screenshot.pics/s/Monosnap_2022-08-22_23-04-14.png)

### Camera 1

![WezTerm Icon - Dark 2 - Camera 1](https://user-images.githubusercontent.com/18397/186060334-df221a69-8098-4ec2-92d7-afb103f3611f.png)

### Camera 2

![WezTerm Icon - Dark 2 - Camera 2](https://user-images.githubusercontent.com/18397/186060362-5acd7906-1894-44d5-accd-18df73d61925.png)

**Comment by gf3** [comment]
Still more tweaking, trying to bring back the colourful playfulness of the light-mode icon:

### ⌘ Tab

![Screenshot](https://screenshot.pics/s/Monosnap_2022-08-23_22-43-08.png)

### Icon

![WezTerm Icon 3](https://user-images.githubusercontent.com/18397/186308089-da43bf63-0524-4eb0-93fc-0829f48096da.png)

**Reply by gf3** [reply]
A little more colourful:

![WezTerm Icon](https://user-images.githubusercontent.com/18397/186718396-1e30e59b-a954-4e21-9af6-8fb401acc026.png)

**Reply by gf3** [reply]
Even more colourful:

![WezTerm Icon More](https://user-images.githubusercontent.com/18397/187488967-d0d61d83-008a-4661-81f0-a75eef7c2a58.png)

**Reply by kttlr** [reply]
I know this an old thread, but did you end up making an .icns for this? I'd love to use it.

**Reply by lycheeje11y** [reply]
Love this! Using it rn. 

@kyle-kettler, you can just download the png file linked above. You might want to use some sketchy online website to convert it to an icns file.

**Reply by emrtnn** [reply]
Love this icon, thank you :D

**Comment by chichid** [comment]
I think it's best if you keep it simple and 2D ideally, to make it cross platform.

**Comment by boltlessengineer** [comment]
Could you make an repository like [kitty-icon](https://github.com/DinkDonk/kitty-icon)?
It would be nice if there is an repo containing Makefile, png, icns files of your icon design.

**Comment by proski** [comment]
Thank you for this effort! The existing icon is barely visible on Windows, Compare that to other icons.
![wezterm-icon](https://user-images.githubusercontent.com/1317472/210019797-aceba15e-6de8-4feb-bd62-7621af5ddcc6.png)

I would prefer a simpler icon that would look good even when very small. I don't want the taskbar to steal too much space from other windows. In fact, changing the background to a light color in the existing icon would make it much easier to discern.

Also, maybe we could remove "$" and just leave a stylish "W"?

**Comment by usagi-flow** [comment]
Some general feedback:

I'd prefer and recommend a 2D icon: FWIW, 2D icons are the current standard, and having an icon that doesn't fit the overall look isn't so nice. A subtle shade in the background would be fine of course.

Regarding the colors, I'd recommend some simplicity. I like the way Apple approaches this: If you look at their icons, for example, you'll notice that most of the time, the background or the foreground consists of a single color. Sometimes they'll use a gradient or rainbow in the foreground, but keep the background in a single color (cf. "Photos" app). This makes the detail come out better.

As for the rounded square shape: 👍

**Comment by netherstar** [comment]
current icon is the best, please don't change it, i love that purple color on black and simple $W text, don't make it look like teenage app!

**Reply by proski** [reply]
Purple on black doesn't have enough contrast to be easily visible. Icons with poor contrast "stand out" as dirty blobs in the middle of the well designed icons. I agree that it's best to keep the icon simple and 2D. Extra borders and other 3D stuff is hard to see if small icons are used.

**Comment by rafo** [comment]
@gf3 I love your icons. Ignore the haters like they should ignore your icons, if they don't like or need them. Having alternatives to the current icon is much needed as the current one has some UI/UX issues. Did you bundle them in a repo?

**Reply by gf3** [reply]
@rafo thank you! i haven't bundled it anywhere as interest seemed to drop off

**Comment by mitchcapper** [comment]
Nice icons!  For windows the titlebar/taskbar icon sizing can be fairly small. Here is a sample of the icon next to some others along with the [ico](https://github.com/wez/wezterm/files/11673894/ico.zip) files.  The 3d icon could be cropped slightly more to  make it somewhat bigger.
![image](https://github.com/wez/wezterm/assets/1643324/552fa05c-9c6e-4868-8481-76bc1505d600)

**Comment by robrecord** [comment]
@gf3 I *really* like your icons - camera 2 for me. I don't like the current WezTerm purple-and-black icon in use, it's really drab. Would you be able to share your lovely icon please?

**Reply by gf3** [reply]
@robrecord thank you, rob. you should be able to just download the png from the above threads :)

**Comment by mikker** [comment]
Love your work @gf3. I went in a slightly more conservative direction with mine. Trying to match Terminal.app and iTerm 2 but also give it some WezTerm feeling.

![icon_1024](https://github.com/wez/wezterm/assets/2819/cf863cb7-8a26-4075-be93-5dc391a05011)

I think, to make it feel as at home as possible, there should be custom icons for each platform. Maybe Windows and Unix can share, not too familiar, but macOS at least has a very distinct style that's unlike Windows'.

**Reply by gf3** [reply]
i love it!

**Reply by chrisgrieser** [reply]
This one fits best in the overall macOS Icon look, in my view <3

**Reply by chrisgrieser** [reply]
two small suggestions:
1. you should upload your icon to https://macosicons.com/#/ – it's more or less the unofficial mac icon repository (e.g., also used by apps like [replacicon](https://replacicon.app/))
2. if you are planing to make alternate versions, I'd prefer the `$W` to be a bit bigger? 🤔

**Reply by mikker** [reply]
Here you go: bigger and a bit more bright https://macosicons.com/#/u/mikker (not approved yet)

**Reply by mikker** [reply]
![icon_1024](https://github.com/wez/wezterm/assets/2819/15956aab-1407-4c57-a0e3-751cee665f07)

**Reply by chrisgrieser** [reply]
thank you <3

**Comment by mikker** [comment]
![CleanShot 2023-06-09 at 20 26 49@2x](https://github.com/wez/wezterm/assets/2819/e6566778-3ff2-446f-8ccf-718a7894d10d)

Next to the others.

@wez are you open to pull requests? Would you like the decide or do you want to do something like a vote? What could be the next steps?

**Reply by wez** [reply]
My stance is:
* I recognize that there are shortcomings in the current icon
* I appreciate all these additional designs
* I don't know that I want to change the default icon, as I don't truly love these alternatives, and the alternatives are often optimized for one platform rather than considering all platforms, and it is important to me for there to be a common and obvious "wezterm mark" on all systems
* I think finding a way to give users the option to choose their preferred icon is more in alignment with the configurability and expressiveness of wezterm than just picking one of these existing options
* That needs to be solved for macOS, Windows and X11/Wayland systems, which all work differently in regard to application icons
* Maybe we just start with folks submitting PRs to `assets/icon` to add additional icons, plus writing up some info on how to change the icon for the different systems?

**Reply by mikker** [reply]
Very reasonable. Can you point me to the part of the setup that handles bundling the apps with icons? Not too familiar with Rust nor Cargo or the distribution process but willing to put some time into making what you described here happen.

**Reply by wez** [reply]
It's kind of involved, but you can find some of the places by grepping:

```console
$ rg assets/icon
README.md
3:<img height="128" alt="WezTerm Icon" src="https://raw.githubusercontent.com/wez/wezterm/main/assets/icon/wezterm-icon.svg" align="left"> *A GPU-accelerated cross-platform terminal emulator and multiplexer written by <a href="https://github.com/wez">@wez</a> and implemented in <a href="https://www.rust-lang.org/">Rust</a>*

wezterm-gui/src/termwindow/mod.rs
92:pub const ICON_DATA: &'static [u8] = include_bytes!("../../../assets/icon/terminal.png");

ci/appimage.sh
12:install -Dm644 assets/icon/terminal.png AppDir/usr/share/icons/hicolor/128x128/apps/org.wezfurlong.wezterm.png

ci/deploy.sh
170:install -Dm644 assets/icon/terminal.png %{buildroot}/usr/share/icons/hicolor/128x128/apps/org.wezfurlong.wezterm.png
244:        install -Dm644 assets/icon/terminal.png pkg/debian/usr/share/icons/hicolor/128x128/apps/org.wezfurlong.wezterm.png
293:  assets/icon/terminal.png
294:  assets/icon/wezterm-icon.svg

ci/build-docs.sh
56:cp "assets/icon/terminal.png" docs/favicon.png
57:cp "assets/icon/wezterm-icon.svg" docs/favicon.svg

ci/generate-workflows.py
12:    "assets/icon/*",

assets/icon/update.sh
5:cd $(git rev-parse --show-toplevel)/assets/icon

assets/flatpak/org.wezfurlong.wezterm.template.json
34:                "install -Dm644 ./assets/icon/terminal.png /app/share/icons/hicolor/128x128/apps/org.wezfurlong.wezterm.png",

assets/flatpak/org.wezfurlong.wezterm.json
38:                "cp ./assets/icon/terminal.png /app/share/icons/hicolor/128x128/apps/org.wezfurlong.wezterm.png",
```

On X11/Wayland: the icon shown in menus is specified via the `.desktop` file and points to the installed org.wezfurlong.wezterm.png file.
On X11: the icon reported via x properties to the window manager is set at runtime from the compiled in ICON_DATA
On macOS: the icon is specified via the Info.plist via the CFBundleIconFile
On Windows: the icon is compiled into the manifest by `wezterm-gui/build.rs` from the file `assets/windows/terminal.ico`.

Most of the packaging logic is found in `ci/deploy.sh`, but there are other scripts run for flatpaks, appimage and windows installers.

**Reply by mikker** [reply]
None of these look close to being dynamically reconfigurable once the thing is built? It would have to be a build flag with a path to a png or something. All three platforms probably have ways to update icons "live". I'll have to investigate.

(Fwiw I'm currently doing this after recompiling WezTerm `cp ~/WezTerm.icns /Applications/WezTerm.app/Contents/Resources/terminal.icns`)

**Reply by wez** [reply]
Yeah, these are baked in at compile time. I'm actively opposed to adding build-time configuration options/feature flags for icons, but I'm open to adjusting the core logic to allow for more runtime flexibility where it makes sense.

I'm not sure what can be done in-app on macOS without rewriting the application plist and potentially invalidating the code signing.

On X11/Wayland, the icon is largely none of wezterm's business, except for the ICON_DATA compiled in resource used on X11. That could turned into a configuration option that specifies a user-provided path or falls back to a compile time option.

On Windows: the icon shown in Explorer comes from the one encoded in the manifest which is a compile time option. Similar to macOS, changing that after build will potentially invalidate any code signing that we do in the future.  The icon shown in the task switcher can probably be overridden at runtime by doing some stuff in the WindowProc, but needs more investigation.

**Comment by proski** [comment]
The only real problem for me is the low contrast between the `$W` and the background. It's out of place for any OS.

Additionally, dark background is somewhat problematic, especially on MacOS. Some MacOS icons have dark background, but they employ gradients around the edge (Apple TV, Stocks) or have a large image filling most of the icon (Clock, Mission Control). Generally, a bright background is a safer bet, especially if a single cross-platform icon is needed.

**Comment by bates64** [comment]
On the contrast problem, @mikker's icon has [a contrast ratio of 2.88 (fail)](https://coolors.co/contrast-checker/1d1d1d-5243fe). This is of course an issue present with the current icon, but we should solve it here too. 

This will be seen as small text, so we would want a contrast ratio of AA or better. For more detail on WCAG 2.0 contrast guidelines, [this is a great, short resource](https://usecontrast.com/guide).

Without changing the background colour, the colour closest to the current choice that meets this is `#7C70FF` ([contrast ratio 4.56](https://coolors.co/contrast-checker/7c70ff-1c1c1c)).

Here's a mockup:

![image](https://github.com/wez/wezterm/assets/9429556/64011092-7dfc-4320-99cd-7a16906e8cf8)

**Comment by cedws** [comment]
Would be cool if the dock icon could be configured (to a built in preset.) Not sure if this is possible.

I am finding that the contrast of the current icon is too low and I find it hard to recognise on the dock.

**Reply by mikker** [reply]
Doesn't sound like it's neither easy to add nor on the roadmap.

I do this `cp ~/WezTerm.icns /Applications/WezTerm.app/Contents/Resources/terminal.icns` and it works.

**Comment by mikker** [comment]
Here's a handy repo if you want to use my logo on macOS: https://github.com/mikker/wezterm-icon

**Comment by cschar** [comment]
Very Helpful!

Here is a script to automate everything on MacOS


Create a new folder with "new_wezterm.png" , and this script locally then run it. 

Then copy output .icns object to 
`/Applications/WezTerm.app/Contents/Resources/terminal.icns"

```bash
#!/bin/bash

#
# Following instructions 
# https://gist.github.com/ansarizafar/6fa64f44aa933794c4d6638eec32b9aa
#

# Original image
INPUT_IMAGE="new_wezterm.png"

# Step 1
echo " ==== Creating icons for $INPUT_IMAGE " 
sips -s format png --resampleHeightWidth 16 16 "${INPUT_IMAGE}" --out "icon_16x16.png"
sips -s format png --resampleHeightWidth 32 32 "${INPUT_IMAGE}" --out "icon_16x16@2x.png"
sips -s format png --resampleHeightWidth 32 32 "${INPUT_IMAGE}" --out "icon_32x32.png"
sips -s format png --resampleHeightWidth 64 64 "${INPUT_IMAGE}" --out "icon_64x64.png"
sips -s format png --resampleHeightWidth 128 128 "${INPUT_IMAGE}" --out "icon_128x128.png"
sips -s format png --resampleHeightWidth 256 256 "${INPUT_IMAGE}" --out "icon_128x128@2x.png"
sips -s format png --resampleHeightWidth 256 256 "${INPUT_IMAGE}" --out "icon_256x256.png"
sips -s format png --resampleHeightWidth 512 512 "${INPUT_IMAGE}" --out "icon_256x256@2x.png"
sips -s format png --resampleHeightWidth 512 512 "${INPUT_IMAGE}" --out "icon_512x512.png"
sips -s format png --resampleHeightWidth 1024 1024 "${INPUT_IMAGE}" --out "icon_512x512@2x.png"

# Step 2
ICONSET_FOLDER="my_wez_icon.iconset"
echo " ==== Moving images to iconset folder"
mkdir $ICONSET_FOLDER
find . -type f \
       -maxdepth 1 \
       -name "*.png" \
       -not -name "${INPUT_IMAGE}" \
       -exec mv {} $ICONSET_FOLDER \;

# Step 3
echo " ==== Converting .iconset folder to icns "
iconutil -c icns $ICONSET_FOLDER
```
<img width="264" alt="Screenshot 2024-05-23 at 8 38 53 PM" src="https://github.com/wez/wezterm/assets/296551/f4aad094-16e7-4d14-a595-c1cb47f3ce06">

**Reply by Risinggoblin** [reply]
qq - why go through with placing the icon inside hte app/contents and no just paste the icon file to "get info" manually? any particular advantages of this approach?+

**Comment by ghost** [comment]
Here's a tutorial in win11 to change the icon. Answers are generated by ChatGPT.
![image](https://github.com/user-attachments/assets/0b4ce0a6-040b-4301-9cc8-6b3d4be6307e)
![image](https://github.com/user-attachments/assets/02dacc40-ab8f-45e4-9794-70a9cb67d3e8)
![image](https://github.com/user-attachments/assets/829f9c03-104c-407d-aa2b-4771b2b7c1fb)

# Prequisites
Download [GIMP](https://www.gimp.org/), [Resource Hacker](https://www.angusj.com/resourcehacker/)

# Steps
## Create .ico File
1. Open GIMP.
2. Go to File > Open and select your PNG file.
3. Go to Image > Scale Image.
4. In the "Scale Image" dialog box, change the image size to the desired resolution (e.g., 256x256, 128x128, 96x96, 64x64, 48x48, 32x32, 16x16).
Click Scale.
5. Go to File > Export As, select PNG format, and save this resized image.
6. Undo the scaling by pressing Ctrl + Z or Edit > Undo Scale Image.
7. Repeat the scaling, exporting, and undoing steps for each desired resolution.
8. Open a new image in GIMP with the largest resolution you need (e.g., 256x256).
9. Go to File > Open As Layers and select all the PNG files you created. Each image will be added as a separate layer.
10. Make sure all layers are visible and in the correct order, with the largest resolution at the top.
11. Go to File > Export As.
12. In the "Export Image" dialog, choose the filename and change the file extension to .ico (e.g., icon.ico).
13. Click Export.
14. In the "Export Image as Windows Icon" dialog, ensure all desired resolutions are checked.
15. Click Export.
![image](https://github.com/user-attachments/assets/2c18b3d1-83d9-4add-a8cb-0ca777882c6d)

## Use Resoucre Hacker to Change Icon
1. Find your `wezterm-gui.exe` location.
2. Launch Resource Hacker and open the `wezterm-gui.exe`.
3. In Resource Hacker, navigate to the "Icon Group" section.
4. Right-click and click "Replace Icon". Then, choose "Open file with new icon" and select the .ico file you just created.
5. Click Replace.
6. Save the changes by clicking "File" > "Save".

**Comment by KaitoMuraoka** [comment]
This discussion was very helpful. Thank you！

I’ve created a script to change the icon on Mac.：
https://github.com/KaitoMuraoka/WeztermChangeIconForMac

Feel free to use it if you’d like. Just place the icon you want to use in the repository, run the script, and restart your system to apply the change.

**Reply by plandem** [reply]
working with OSX with turned off SIP only

**Comment by rcarmo** [comment]
I really like the macOS alternatives here. The default icon, although... er... cross-platform, is just not aesthetically pleasing enough on macOS, so I hope @mikker's icon is adopted there.

**Reply by plandem** [reply]
[`fileicon`](https://github.com/mklement0/fileicon) can be used to apply any icon for OSX apps.

**Comment by grapao** [comment]
I also made a few icons, I thought that the previous ones were either not colorful enough or not flat enough so here is my take 
<img width="759" alt="image" src="https://github.com/user-attachments/assets/239539d2-4fd8-44b6-bf82-6ee462cfea6d" />

To get any of them here is the figma link, you can also modify the gradients as you please: https://www.figma.com/community/file/1486670726006709753
To change the gradients: go in the left menu in Layers -> choose any of them to modify -> Fond -> Artwork -> Icon Center. Click on it and in the right menu scroll down and you should see a Fill section with a linear gradient. 
To export click on the asset you want and scroll down on the right menu again and there should be an Export section

**Comment by jasonm23** [comment]
Hi all, I'm setting up an alternate Windows build repo that'll use some of the icons here, since using `rcedit.exe` didn't seem to work correctly when I attempted to patch `wezterm-gui.exe`

The repo is: https://github.com/ocodo/wezterm-alt-windows-icon-builds

Credit has been added to your icons @gf3 @grapao @mikker

**Comment by jasonm23** [comment]
Just a quick heads up, after fighting with cargo build most of yesterday, I then figured out Windows icon caching can be a bit ... of a challenge.

Turns out `rcedit` patching `wezterm-gui.exe` was all that's needed. Just need to flush the icon cache restart Windows. (note: killing `explorer.exe`, deleting `iconcache*` files, starting `explorer.exe` partially works, but has some hangers on)

Anyway I'll be pushing the patched releases "today" (UTC+8:00 here).

@gf3 I'll also add your 3D Key style icons.

**Comment by jasonm23** [comment]
There we go: 

- https://github.com/ocodo/wezterm-alt-windows-icon-builds/releases

**Comment by jasonm23** [comment]
@gf3 just in the process of adding your 3D key icons now they can be renamed, I just made some names up for wip.

Will be up later, just testing my workflows on a test repo first

**Comment by jasonm23** [comment]
Ok, all automated now on https://github.com/ocodo/wezterm-alt-windows-icon-builds

So it'll build / update the release if we add new icons.

~~I'm going to spin up a similar project that releases "unpadded" versions of the pngs as macOS `.icns` too.~~ @KaitoMuraoka's project covers that I think.

**Reply by noidilin** [reply]
Thanks for these windows builds.

I have generated app manifest for all alt icon builds to manage them with scoop. 
They can be used as the template: [my scoop bucket](https://github.com/noidilin/scoop/tree/master/bucket).

Note that the icon 'mikker-black' is set as default 'wezterm' manifest in favor of my personal preference.

**Reply by jasonm23** [reply]
You are welcome to add the installation notes to the project as a PR

**Reply by noidilin** [reply]
sure, I am willing to work on this next week, once I have time.

**Reply by noidilin** [reply]
I've submitted a PR to [ocodo/wezterm-alt-windows-icon-builds](https://github.com/ocodo/wezterm-alt-windows-icon-builds). 
Feel free to modify the contents in it.

**Comment by jasonm23** [comment]
oh, if you want me to change the icon names @gf3 @mikker @grapao - let me know.

**Reply by gf3** [reply]
looks great—thank you!
